### ISSUE TITLE
Q: Why is `tfp.optimizer.lbfgs_minimize` a pure function?

### ISSUE THREAD

**Issue by joelberkeley** [tensorflow/probability#1278]
I'm interested from a design perspective why `tfp.optimizer.lbfgs_minimize` is a pure function (i.e. that it doesn't modify its arguments), unlike, say `tf.optimizers.Adam.minimize` or indeed `scipy.optimize.minimize`?

BTW I like it being pure, better for FP design.

**Comment by brianwa84** [comment]
Pure functions allow us to be tensor-in, tensor-out, as opposed to
requiring users to use tf.Variable, which works with TF but not with JAX.
In TF1, Variable introduced a fair amount of performance overhead. While
this is much improved with TF2, we then added the JAX consideration.
In contrast to scipy, TF Tensor objects are not modifiable, unlike
np.array.
In contrast to tf.opt.Adam, we want to operate on pure functions instead of
implicitly defined loss(batch|NN weight).
Incidentally, you might notice that jax has a pure adam optimizer
(jax.experimental.optimizers.adam).

Brian Patton | Software Engineer | ***@***.***

**Comment by joelberkeley** [comment]
@brianwa84 thanks, that helps
